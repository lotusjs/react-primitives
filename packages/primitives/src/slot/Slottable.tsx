import React, { isValidElement } from 'react';

export function Slottable({ children }: { children: React.ReactNode }) {
  return <>{children}</>;
}

export function isSlottable(child: React.ReactNode): child is React.ReactElement {
  return isValidElement(child) && child.type === Slottable;
}
