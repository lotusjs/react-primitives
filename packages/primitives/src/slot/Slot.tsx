import React, { Children, cloneElement, forwardRef, isValidElement } from 'react';
import { isSlottable } from './Slottable';
import { SlotClone } from './SlotClone';

export interface SlotProps extends React.HTMLAttributes<HTMLElement> {
  children?: React.ReactNode;
}

export const Slot = forwardRef<HTMLElement, SlotProps>(
  (props, forwardedRef) => {
    const { children, ...slotProps } = props;
    // eslint-disable-next-line react/no-children-to-array
    const childrenArray = Children.toArray(children);
    const slottable = childrenArray.find(isSlottable);

    if (slottable) {
      const newElement = slottable.props.children as React.ReactNode;

      const newChildren = childrenArray.map((child) => {
        if (child === slottable) {
          // eslint-disable-next-line react/no-children-count
          if (Children.count(newElement) > 1)
            // eslint-disable-next-line react/no-children-only
            return Children.only(null);

          return isValidElement(newElement)
            ? (newElement.props.children as React.ReactNode)
            : null;
        }
        else {
          return child;
        }
      });

      return (
        <SlotClone {...slotProps} ref={forwardedRef}>
          {isValidElement(newElement)
            // eslint-disable-next-line react/no-clone-element
            ? cloneElement(newElement, undefined, newChildren)
            : null}
        </SlotClone>
      );
    }

    return (
      <SlotClone {...slotProps} ref={forwardedRef}>
        {children}
      </SlotClone>
    );
  },
);

Slot.displayName = 'Slot';
