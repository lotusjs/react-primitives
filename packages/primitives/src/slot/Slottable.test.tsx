import React from 'react';
import { describe, expect, it } from 'vitest';
import { render } from '@testing-library/react';
import { Slot } from './Slot';
import { Slottable } from './Slottable';

const Button = React.forwardRef<
  React.ElementRef<'button'>,
  React.ComponentProps<'button'> & {
    asChild?: boolean;
    iconLeft?: React.ReactNode;
    iconRight?: React.ReactNode;
  }
>(({ children, asChild = false, iconLeft, iconRight, ...props }, forwardedRef) => {
  const Comp = asChild ? Slot : 'button';
  return (
    <Comp {...props} ref={forwardedRef}>
      {iconLeft}
      <Slottable>{children}</Slottable>
      {iconRight}
    </Comp>
  );
});

describe('given a Button with Slottable', () => {
  describe('without asChild', () => {
    it('should render a button with icon on the left/right', async () => {
      const tree = render(
        <Button iconLeft={<span>left</span>} iconRight={<span>right</span>}>
          Button
          {' '}
          <em>text</em>
        </Button>,
      );

      expect(tree.container).toMatchSnapshot();
    });

    describe('with asChild', () => {
      it('should render a link with icon on the left/right', async () => {
        const tree = render(
          <Button iconLeft={<span>left</span>} iconRight={<span>right</span>} asChild>
            <a href="https://radix-ui.com">
              Button
              {' '}
              <em>text</em>
            </a>
          </Button>,
        );

        expect(tree.container).toMatchSnapshot();
      });
    });
  });
});
