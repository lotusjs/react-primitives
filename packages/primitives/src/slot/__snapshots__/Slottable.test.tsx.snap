// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`given a Button with Slottable > without asChild > should render a button with icon on the left/right 1`] = `
<div>
  <button>
    <span>
      left
    </span>
    Button
     
    <em>
      text
    </em>
    <span>
      right
    </span>
  </button>
</div>
`;

exports[`given a Button with Slottable > without asChild > with asChild > should render a link with icon on the left/right 1`] = `
<div>
  <a
    href="https://radix-ui.com"
  >
    <span>
      left
    </span>
    Button
     
    <em>
      text
    </em>
    <span>
      right
    </span>
  </a>
</div>
`;
